src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs

```diff
@@ -0,0 +1,177 @@
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Mvc.Internal
+{
+    public class ApiBehaviorApplicationModelProvider : IApplicationModelProvider
+    {
+        private readonly ApiBehaviorOptions _apiBehaviorOptions;
+        private readonly ModelMetadataProvider _modelMetadataProvider;
+        private readonly ModelStateInvalidFilter _modelStateInvalidFilter;
+        private readonly ILogger _logger;
+
+        public ApiBehaviorApplicationModelProvider(
+            IOptions<ApiBehaviorOptions> apiBehaviorOptions,
+            IModelMetadataProvider modelMetadataProvider,
+            ILoggerFactory loggerFactory)
+        {
+            _apiBehaviorOptions = apiBehaviorOptions.Value;
+            _modelMetadataProvider = modelMetadataProvider as ModelMetadataProvider;
+            _logger = loggerFactory.CreateLogger<ApiBehaviorApplicationModelProvider>();
+
+            if (_apiBehaviorOptions.EnableModelStateInvalidFilter && _apiBehaviorOptions.InvalidModelStateResponseFactory == null)
+            {
+                throw new ArgumentException(Resources.FormatPropertyOfTypeCannotBeNull(
+                    typeof(ApiBehaviorOptions),
+                    nameof(ApiBehaviorOptions.InvalidModelStateResponseFactory)));
+            }
+
+            _modelStateInvalidFilter = new ModelStateInvalidFilter(
+                apiBehaviorOptions.Value,
+                loggerFactory.CreateLogger<ModelStateInvalidFilter>());
+        }
+
+        /// <remarks>
+        /// Order is set to execute after the <see cref="DefaultApplicationModelProvider"/> and allow any other user
+        /// <see cref="IApplicationModelProvider"/> that configure routing to execute.
+        /// </remarks>
+        public int Order => -1000 + 100;
+
+        public void OnProvidersExecuted(ApplicationModelProviderContext context)
+        {
+        }
+
+        public void OnProvidersExecuting(ApplicationModelProviderContext context)
+        {
+            foreach (var controllerModel in context.Result.Controllers)
+            {
+                var isApiController = controllerModel.Attributes.OfType<IApiBehaviorMetadata>().Any();
+                var controllerHasSelectorModel = controllerModel.Selectors.Any(s => s.AttributeRouteModel != null);
+
+                foreach (var actionModel in controllerModel.Actions)
+                {
+                    if (!isApiController && !actionModel.Attributes.OfType<IApiBehaviorMetadata>().Any())
+                    {
+                        continue;
+                    }
+
+                    if (!controllerHasSelectorModel && !actionModel.Selectors.Any(s => s.AttributeRouteModel != null))
+                    {
+                        // Require attribute routing with controllers annotated with ApiControllerAttribute
+                        throw new InvalidOperationException(Resources.FormatApiController_AttributeRouteRequired(nameof(ApiControllerAttribute)));
+                    }
```

Comment:
Can we separate all of these 'logical operations' into private methods? I feel like this list of steps is going to grow and it's already at like 3.